### PR TITLE
feat: Squares of Residuals option (PT-181940029)

### PR DESCRIPTION
**Pull request by emcelroy**
PT Story: https://www.pivotaltracker.com/story/show/181940029

### CODE REVIEW

File: v3/src/components/graph/components/scatter-plot-utils.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    const lineYCoord = yScale(slope * xValue + intercept) / numExtraSecondaryBands + rightCoord
+    const residualCoord = yCoord - lineYCoord
+    const lineXCoord = xCoord + residualCoord
+    const x = Math.min(xCoord, lineXCoord)
+    const y = Math.min(yCoord, lineYCoord)
+    const side = Math.abs(residualCoord)
+    return { caseID, color, side, x, y }
+  }
+
+  function residualSquares(lineDescriptions: ILineDescription[]) {
```
Comment (emcelroy) [inline v3/src/components/graph/components/scatter-plot-utils.ts:59]: I'm having second thoughts about this function name. It's very similar to the `residualSquare` function name and may be confusing in that regard. Would it be better to name the `residualSquare` function something like `residualSquareAttributes` or `residualSquareDescriptions`? Or is there a better name for the `residualSquares` function?

Comment (emcelroy) [reply v3/src/components/graph/components/scatter-plot-utils.ts:59]: Those sound good to me 👍

File: v3/cypress/e2e/adornments.spec.ts [hunk 2]
```diff
@@ -532,4 +532,44 @@ context("Graph adornments", () => {
     // cy.get(".movable-value-label").should("have.length", 0)
     // cy.get(".movable-value-fill").should("have.length", 0)
   })
+  it.only("adds squares of residuals squares to the plot when the Squares of Residuals checkbox is checked", () => {
```
Comment (kswenson) [inline v3/cypress/e2e/adornments.spec.ts:535]:
```suggestion
  it("adds squares of residuals squares to the plot when the Squares of Residuals checkbox is checked", () => {
```

File: v3/src/components/graph/adornments/lsrl/lsrl-adornment-model.ts [hunk 3]
```diff
@@ -75,6 +104,18 @@ export const LSRLAdornmentModel = AdornmentModel
       }
     })
     return caseValues
+  },
+  get lineDescriptions() {
+    const lineDescriptions: ILineDescription[] = []
+    self.lines.forEach((linesArray, key) => {
+      linesArray.forEach(line => {
+        const { category, intercept, slope } = line
+        if (!intercept || !isFinite(intercept) || !slope || !isFinite(slope)) return
```
Comment (kswenson) [inline v3/src/components/graph/adornments/lsrl/lsrl-adornment-model.ts:113]:
```suggestion
        if (!Number.isFinite(intercept) || !Number.isFinite(slope)) return
```
`intercept` and `slope` can be zero for valid lines.

File: v3/src/components/graph/components/scatter-plot-utils.ts (hunk 1 above)
Comment (kswenson) [reply v3/src/components/graph/components/scatter-plot-utils.ts:59]: Were you going to rename the second function?

File: v3/src/components/graph/utilities/graph-utils.ts [hunk 4]
```diff
@@ -157,16 +157,22 @@ export function lineToAxisIntercepts(iSlope: number, iIntercept: number,
   }
 }
 
-export function equationString(slope: number, intercept: number, attrNames: {x: string, y: string}) {
-  const float = format('.4~r')
+export function equationString(
+  slope: number, intercept: number, attrNames: {x: string, y: string}, sumOfSquares?: number
+) {
```
Comment (kswenson) [inline v3/src/components/graph/utilities/graph-utils.ts:162]:
```suggestion
interface IEquationString {
  slope: number
  intercept: number
  attrNames: {x: string, y: string}
  sumOfSquares?: number
}
export function equationString({ slope, intercept, attrNames, sumOfSquares }: IEquationString) {
```

File: v3/src/components/graph/utilities/graph-utils.ts [hunk 5]
```diff
@@ -157,16 +157,22 @@ export function lineToAxisIntercepts(iSlope: number, iIntercept: number,
...
+  const float = format(".3~r")
+  const floatSumOfSquares = format(",.0f")
```
Comment (kswenson) [inline v3/src/components/graph/utilities/graph-utils.ts:164]: Is this what v2 uses? Seems like there could be data sets with very small residuals where this wouldn't work well.

File: v3/src/components/graph/utilities/graph-utils.ts [hunk 6]
```diff
@@ -157,16 +157,22 @@ export function lineToAxisIntercepts(iSlope: number, iIntercept: number,
...
   if (isFinite(slope) && slope !== 0) {
     const xAttrString = attrNames.x.length > 1 ? `(<em>${attrNames.x}</em>)` : `<em>${attrNames.x}</em>`
     const interceptString = intercept === 0
       ? ""
       : intercept > 0
         ? ` + ${float(intercept)}`
         : ` ${float(intercept)}`
-    return `<em>${attrNames.y}</em> = ${float(slope)} ${xAttrString}${interceptString}`
+    const squaresPart = sumOfSquares || sumOfSquares === 0
```
Comment (kswenson) [inline v3/src/components/graph/utilities/graph-utils.ts:172]:
```suggestion
    const squaresPart = Number.isFinite(sumOfSquares)
```

File: v3/src/components/graph/utilities/graph-utils.ts [hunk 7]
```diff
@@ -187,12 +193,13 @@ export function percentString(value: number) {
 
 export const lsrlEquationString = (
   slope: number, intercept: number, attrNames: {x: string, y: string}, caseValues: Point[],
-  rSquared?: number, confidenceBandsEnabled?: boolean, color?: string, interceptLocked=false
+  confidenceBandsEnabled?: boolean, rSquared?: number, color?: string, interceptLocked=false, sumOfSquares?: number
```
Comment (kswenson) [inline v3/src/components/graph/utilities/graph-utils.ts:196]: Another case where switching to passing a structure would be an improvement.

File: v3/src/components/graph/utilities/graph-utils.ts [hunk 8]
```diff
@@ -207,9 +214,12 @@ export const lsrlEquationString = (
   const seSlopePart = confidenceBandsEnabled && !interceptLocked
     ? `<br />SE<sub>slope</sub> = ${floatSeSlope(seSlope)}`
     : ""
+  const squaresPart = sumOfSquares || sumOfSquares === 0
```
Comment (kswenson) [inline v3/src/components/graph/utilities/graph-utils.ts:217]:
```suggestion
  const squaresPart = Number.isFinite(sumOfSquares)
```

File: v3/src/components/graph/utilities/graph-utils.ts [hunk 9]
```diff
@@ -187,12 +193,13 @@ export function percentString(value: number) {
...
 export const lsrlEquationString = (
   slope: number, intercept: number, attrNames: {x: string, y: string}, caseValues: Point[],
-  rSquared?: number, confidenceBandsEnabled?: boolean, color?: string, interceptLocked=false
+  confidenceBandsEnabled?: boolean, rSquared?: number, color?: string, interceptLocked=false, sumOfSquares?: number
 ) => {
   const float = format(".3~r")
   const floatIntercept = format(".1~f")
   const floatSeSlope = format(".3~f")
   const linearRegression = leastSquaresLinearRegression(caseValues, interceptLocked)
+  const floatSumOfSquares = format(",.0f")
```
Comment (kswenson) [inline v3/src/components/graph/utilities/graph-utils.ts:202]: Use named constants for these format specifiers so if we need to change them they can be changed in one place. I'm concerned that these fixed-decimal place formats won't work well for some data sets. How does v2 handle formatting these values?

File: v3/src/components/graph/adornments/movable-line/movable-line-adornment-model.ts [hunk 10]
```diff
@@ -24,6 +57,12 @@ export const MovableLineInstance = types.model("MovableLineInstance", {
   },
   setPivot2(point: Point) {
     self.pivot2.set(point)
+  },
+  setVolatileIntercept(intercept?: number) {
+    self.dynamicIntercept = intercept
+  },
+  setVolatileSlope(slope?: number) {
+    self.dynamicSlope = slope
```
Comment (kswenson) [inline v3/src/components/graph/adornments/movable-line/movable-line-adornment-model.ts:65]: You don't need two functions given that they're always called together:
```suggestion
  setVolatile(intercept?: number, slope?: number) {
    self.dynamicIntercept = intercept
    self.dynamicSlope = slope
```

File: v3/src/components/graph/adornments/movable-line/movable-line-adornment-model.ts [hunk 11]
```diff
@@ -33,14 +72,33 @@ export const MovableLineAdornmentModel = AdornmentModel
   type: types.optional(types.literal(kMovableLineType), kMovableLineType),
   lines: types.map(MovableLineInstance)
 })
+.views(self => ({
+  get lineDescriptions() {
+    const lineDescriptions: ILineDescription[] = []
+    self.lines.forEach((line, key) => {
+      const { intercept, slope } = line.slopeAndIntercept
+      if (!isFinite(intercept) || !isFinite(slope)) return
```
Comment (kswenson) [inline v3/src/components/graph/adornments/movable-line/movable-line-adornment-model.ts:80]:
```suggestion
      if (!Number.isFinite(intercept) || !Number.isFinite(slope)) return
```

File: v3/src/components/graph/adornments/movable-line/movable-line-adornment-model.ts [hunk 12]
```diff
@@ -33,14 +72,33 @@ export const MovableLineAdornmentModel = AdornmentModel
...
   setLine(
-    aLine: {intercept: number, slope: number, pivot1?: Point, pivot2?: Point, equationCoords?: Point}, key=''
+    aLine: {intercept: number, slope: number, pivot1?: Point, pivot2?: Point, equationCoords?: Point}, key=""
   ) {
     self.lines.set(key, aLine)
     const line = self.lines.get(key)
     line?.setPivot1(aLine.pivot1 ?? kInfinitePoint)
     line?.setPivot2(aLine.pivot2 ?? kInfinitePoint)
+  },
+  updateVolatileProps(
```
Comment (kswenson) [inline v3/src/components/graph/adornments/movable-line/movable-line-adornment-model.ts:96]:
```suggestion
  setVolatileLine(
```
I realize the implementation is somewhat different, but they're doing the same thing from the client's perspective.

File: v3/src/components/graph/adornments/movable-line/movable-line-adornment-component.tsx [hunk 13]
```diff
@@ -213,9 +216,16 @@ export const MovableLineAdornment = observer(function MovableLineAdornment(props
...
     refreshEquation(slope, intercept)
 
+    // Until the user releases the line, only update the model's volatile props for the slope and intercept. Once
+    // the user releases the line, update the model's slope and intercept and set the volatile props to undefined.
+    // We don't want to save the values with every move of the line, but we do need to make the current values 
+    // available to other clients of the model via the volatile props.
     if (isFinished) {
       const equationCoords = lineParams?.equationCoords
       model.setLine({slope, intercept: newIntercept, equationCoords}, instanceKey)
+      model.updateVolatileProps({intercept: undefined, slope: undefined}, instanceKey)
```
Comment (kswenson) [inline v3/src/components/graph/adornments/movable-line/movable-line-adornment-component.tsx:226]:
```suggestion
```
`setLine` should handle this internally.

File: v3/src/components/graph/adornments/movable-line/movable-line-adornment-component.tsx [hunk 14]
```diff
@@ -297,6 +311,9 @@ export const MovableLineAdornment = observer(function MovableLineAdornment(props
           },
           instanceKey
         )
+        model.updateVolatileProps({intercept: undefined, slope: undefined}, instanceKey)
```
Comment (kswenson) [inline v3/src/components/graph/adornments/movable-line/movable-line-adornment-component.tsx:314]:
```suggestion
```
`setLine` should handle this internally.

**Review by kswenson — CHANGES_REQUESTED** [review]
👍 Looks good -- a bunch of mostly minor/local suggestions.

**Review by bfinzer — APPROVED** [review]
👍🏻LGTM

File: v3/src/components/graph/utilities/graph-utils.ts (hunk 5 above)
Comment (emcelroy) [reply v3/src/components/graph/utilities/graph-utils.ts:170]: @kswenson This isn't exactly what V2 uses. The use of D3's `format` was something of a placeholder when it was first introduced in graph-utils.ts. I looked into this a bit and discussed it with @bfinzer. We determined it'd be best to make matching V2's formatting a separate piece of work. I added [a story about that](https://www.pivotaltracker.com/story/show/186611743) to PT.

File: v3/src/components/graph/adornments/lsrl/lsrl-adornment-component.tsx [hunk 15]
```diff
@@ -149,12 +150,16 @@ export const LSRLAdornment = observer(function LSRLAdornment(props: IProps) {
...
       const screenX = xScale((pointsOnAxes.current.pt1.x + pointsOnAxes.current.pt2.x) / 2) / xSubAxesCount
       const screenY = yScale((pointsOnAxes.current.pt1.y + pointsOnAxes.current.pt2.y) / 2) / ySubAxesCount
       const attrNames = {x: xAttrName, y: yAttrName}
-      const string = lsrlEquationString(
-        slope, intercept, attrNames, caseValues, rSquared, showConfidenceBands, catColor, interceptLocked
-      )
+      const equationParams = {
+        attrNames, caseValues, catColor, intercept, interceptLocked, rSquared, showConfidenceBands, slope, sumOfSquares
+      }
+      const string = lsrlEquationString(equationParams)
```
Comment (kswenson) [inline v3/src/components/graph/adornments/lsrl/lsrl-adornment-component.tsx:162]: You could just use:
```suggestion
      const string = lsrlEquationString({
        attrNames, caseValues, catColor, intercept, interceptLocked, rSquared, showConfidenceBands, slope, sumOfSquares
      })
```

File: v3/src/components/graph/adornments/lsrl/lsrl-adornment-model.ts [hunk 16]
```diff
@@ -23,6 +26,33 @@ export const LSRLInstance = types.model("LSRLInstance", {
...
+    const dataset = dataConfig?.dataset
+    const caseData = dataset?.cases
+    const xAttrID = dataConfig?.attributeID("x") ?? ""
+    const yAttrID = dataConfig?.attributeID("y") ?? ""
+    let sumOfSquares = 0
+    caseData?.forEach((datum: any) => {
+      const fullCaseData = dataConfig?.dataset?.getCase(datum.__id__)
+      if (fullCaseData && dataConfig?.isCaseInSubPlot(cellKey, fullCaseData)) {
+        const x = dataset?.getNumeric(datum.__id__, xAttrID) ?? 0
+        const y = dataset?.getNumeric(datum.__id__, yAttrID) ?? 0
```
Comment (kswenson) [inline v3/src/components/graph/adornments/lsrl/lsrl-adornment-model.ts:40]: Defaulting to zero here seems dangerous. It shouldn't ever happen, but if it does, 0 will invalidate the result. I think `NaN` would be preferable.

File: v3/src/components/graph/adornments/movable-line/movable-line-adornment-model.ts [hunk 17]
```diff
@@ -5,16 +5,49 @@ import { AdornmentModel, IAdornmentModel, IUpdateCategoriesOptions, PointModel,
...
+        const lineY = slope * x + intercept
+        const residual = y - lineY
+        if (isFinite(residual)) {
+          sumOfSquares += residual * residual
+        }
+      }
+    })
+    return sumOfSquares
+  }
+}))
```
Comment (kswenson) [inline v3/src/components/graph/adornments/movable-line/movable-line-adornment-model.ts:50]: Given the similarity between this and the corresponding LSRL version, can it be moved to a utility function, where the defaulting to 0 vs. `NaN` issue can be resolved in one place?

File: v3/src/components/graph/components/scatter-plot-utils.ts [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+          const square = residualSquare(slope, intercept, caseData.__id__)
+          if (!isFinite(square.x) || !isFinite(square.y)) return
+          squares.push(square)
+        }
+      })
+    })
+    return squares
+  }
+
+  return { getXCoord, getYCoord, getCaseCoords, residualSquare, residualSquares: residualSquaresForLines }
```
Comment (kswenson) [inline v3/src/components/graph/components/scatter-plot-utils.ts:82]:
```suggestion
  return { getXCoord, getYCoord, getCaseCoords, residualSquare, residualSquaresForLines }
```
Clients may as well use the more descriptive name as well.

**Review by kswenson — CHANGES_REQUESTED** [review]
👍 Looks good -- just a few final tweaks to suggest.

File: v3/src/components/graph/utilities/graph-utils.ts [hunk 19]
```diff
@@ -494,6 +495,36 @@ export const leastSquaresLinearRegression = (iValues: Point[], iInterceptLocked:
...
+  const dataset = dataConfig?.dataset
+  const caseData = dataset?.cases
+  const xAttrID = dataConfig?.attributeID("x") ?? ""
+  const yAttrID = dataConfig?.attributeID("y") ?? ""
+  let sumOfSquares = 0
+  caseData?.forEach((datum: any) => {
+    const fullCaseData = dataConfig?.dataset?.getCase(datum.__id__)
+    if (fullCaseData && dataConfig?.isCaseInSubPlot(cellKey, fullCaseData)) {
+      const x = dataset?.getNumeric(datum.__id__, xAttrID) ?? defaultVal
+      const y = dataset?.getNumeric(datum.__id__, yAttrID) ?? defaultVal
```
Comment (kswenson) [inline v3/src/components/graph/utilities/graph-utils.ts:516]: I would think that `NaN` is always the preferable default so you wouldn't need the argument.

**Review by kswenson — APPROVED** [review]
👍 Changes look good -- just one remaining question/suggestion.

**Pull request merged**